web/html/src/manager/minion/ansible/accordion-path-content.tsx

```diff
@@ -0,0 +1,182 @@
+        const playbookObj = content[playbookName];
+        const playbookObjKeys = Object.keys(playbookObj);
+        const newPlaybookObject: PlaybookDetails = {
+          name: playbookName,
+          fullPath: playbookObjKeys.includes("fullpath") ? playbookObj["fullpath"] : "-",
+          path: this.props.path,
+          customInventory: playbookObjKeys.includes("custom_inventory") ? playbookObj["custom_inventory"] : "-"
+        }
+        playbookDetailsList.push(newPlaybookObject);
+      }
+    );
+
+    return playbookDetailsList;
+  }
+
+  renderPlaybookPathContent() {
+    if (this.state.content === null || this.state.content.length === 0) {
+      return <div>{t("No Playbook found.")}</div>
+    }
+
+    const content: PlaybookDetails[] = this.state.content;
+    return content.map((p, i) =>
+      <div key={p.toString() + "_" + i.toString()}>
+        { i === 0 ? <br/> : null }
+        <dl className="row">
+          <dt className="col-xs-2">{t('Playbook File Name')}:</dt>
+          <dd className="col-xs-8">{p.name}</dd>
+        </dl>
+        <dl className="row">
+          <dt className="col-xs-2">{t('Full Path')}:</dt>
+          <dd className="col-xs-8">{p.fullPath}</dd>
+        </dl>
+        <dl className="row">
+          <dt className="col-xs-2">{t('Custom Inventory')}:</dt>
+          <dd className="col-xs-8">{p.customInventory}</dd>
+        </dl>
+        { i < content.length -1 ? <hr/> : null }
+      </div>
+    );
+  }
+
+  digestInventoryPathContent(blob: any) {
+    return blob
+  }
+
+  renderInventoryPathContent() {
+    if (this.state.content === null || this.state.content.length === 0) {
+      return <div>{t("Inventory file not found or empty.")}</div>
+    }
+
+    return <AceEditor
+              className="form-control"
+              id="content-state"
+              minLines={20}
+              maxLines={40}
+              readOnly={true}
+              mode="yaml"
+              content={this.state.content}
+            ></AceEditor>;
+  }
+
+  render() {
+    const header =
+      <div className="panel-heading pointer" onClick={() => this.onToggle()}>
+        <h6>
+          <i className={this.state.open || this.state.loading ? "fa fa-chevron-down" : "fa fa-chevron-right"} />
+          { this.props.path.path }
+        </h6>
+      </div>;
+
+    const errors = this.state.errors.length > 0 ? <Messages items={Utils.error(this.state.errors)} /> : null;
+    return (
+      <div className="panel panel-default">
+        {header}
+        <div>
+          {
+            this.state.loading?
+              <Loading text={t("Loading content..")} />
+              :
+              this.state.open ?
+                <>
+                  {
+                    this.state.errors.length > 0 ?
+                      errors
+                      :
+                      this.props.path.type === "playbook" ? this.renderPlaybookPathContent() : this.renderInventoryPathContent()
```

Comment:
It's been quite a while since I've seen this many nested ternaries. 😄 I don't mind though, this works and I don't really have a better recommendation. If you want to refactor this in the future you could move some of it into a standalone function where you can return early and then have fewer branches. This works as is though.

Reply:
+1ed

I find the branching in the component quite complex. I think this component has maybe too much responsibilities (as it takes care of both inventories and playbooks in very specific way. To be concrete: `if (type=='playbook') do something else do something else` is present multiple times in the component.

I'd have some suggestion for improvement (not sure if it's react-ish enough): I think the branching could be avoided if some pieces of logic was injected from outside (namely the `digestXYZ` and `renderXYZ`functions.). In the end, we would use the component like `<AccordionPathContent digestFn=digestPlaybook renderFn=renderPlaybook />`, alternatively in one piece: `<AccordionPathContent contentProcessor=someComponentThatKnowsHowToDigestAndRenderPlaybook />`. Then the accordion component wouldn't care, if it's rendering a playbook or an inventory. Then the user of the component would decide, which logic to inject and the code in the `AccordionPathContent` should be cleaner, in theory. If this sounds confusing, I can try to come up with a patch.

Even if you think the above is a good idea, I don't insist on implementing this as part of the PR (the schedule is very tight), but I'd like to create an issue that would be addressed after the SUMA release deadline.

Reply:
:heart: @hustodemon because you make me feel I was on the right mood at the beginning. I started willing to have a completely ignorant Accordion component without taking care of the content, but (my bad and I am the one to blame) because of time pressure and being in a rush I didn't wanted to spend time thinking about it with the perfect approach and, instead, I didn't wanted to come up with something in the middle that needs to be refactored anyway, so I went for the fast-line implementation, trying just to keep it refactorable.

Initial idea was to have the Accordion to render given children passed by the parent, but I'd like also to come up with an Accordion component able to just do the open/close behavior and also  the loading part. I mean, it would be nice to pass an URL to the Accordion component if it has to load the content once it is called to open, but then it comes the management with the children to render.

@Etheryte : I actually don't like it that much as well, but I tend (not always succeeding though) to extract what has business logic or reusable piece in separate functions, and try to keep everything that is only related to rendering in the `render` method even with some condition (if it does not lead to an unreadable chunk). This is a very borderline case, and I agree it could be written better. Indeed Ithe idea was I extracted the shaping of the content in `renderPlaybookPathContent()` and `renderInventoryPathContent, right because I didn't want to put the two completely different output one next the other, but still the two methods are only rendering pieces, so the ternary condition is visible I can realize the output shape depends on that forking path.

Nevertheless I would love a more readable code (extracting reusable or non context related parts, shorter and maintainable components, ignorant context related components, etc), and I think it would be nice if we can think and plan something about it for the whole codebase (on the React side I mean). @Etheryte would you be interested to take part into some task like this?